### ISSUE TITLE
TestNetworkTransport

### ISSUE THREAD

**Issue by martinbonnin** [apollographql/apollo-kotlin#3737]
Now that we have [test builders](https://www.apollographql.com/docs/kotlin/advanced/test-builders/), we can use them to make writing tests easier. The testing can happen at 2 levels:

1. HTTP
2. GraphQL (this issue)

The `NetworkTransport` interface is relatively simple:

```kotlin
interface NetworkTransport {

  fun <D : Operation.Data> execute(
      request: ApolloRequest<D>,
  ): Flow<ApolloResponse<D>>

  fun dispose()
}
```

`TestNetworkTransport` can maintain a `MutableMap<Operation, ApolloResponse>` so that it responds with the pre-recorded `ApolloResponse` for the given operation (that would be build from the test builders).

Building a test client:

```kotlin
val testApolloClient = ApolloClient.Builder()
       .networkTransport(TestNetworkTransport())
       .build()
```

Ideally, the API would look like:

```kotlin
val testNetworkTransport = testApolloClient.testNetworkTransport // this is an extension function that throws if the networkTransport is not a TestNetworkTransport

val query = HeroQuery()
testNetworkTransport.registerResponse(query) {
  hero = humanHero {
    name = "Luke"
  }
}

// Use it
testApolloClient.query(query).execute().data?.hero?.name 
// "Luke" 
```

The `registerOperation {}` part is the hard part because there's no simple generic way to get the `HeroQuery_TestBuilder` from a `HeroQuery` reified type parameter. We might be able to use some compiler plugin magic like [kotlinx.serialization](https://github.com/Kotlin/kotlinx.serialization/issues/1348) but that adds a new component into the equation.

Another solution might be to have a global registry of `_TestBuilder` keyed by `KClass` but even there I'm not sure how to retrieve them without reflection, which sounds like it's doable on the JVM but not always desired and not possible at all on MPP

**Comment by BoD** [comment]
Just a thought: for the logic returning the `ApolloResponse`, in addition to a "map oriented" API it may be useful to also have a "queue oriented" one and/or one where you can pass a lambda for custom logic.  (I'm thinking about a test scenario where the same query executed twice returns different responses). Something like:

```kotlin
var count = 0
testNetworkTransport.register(query) { query ->
  count++
  when (count) {
    1 -> {
      hero = humanHero {
        name = "Luke"
      }
    }
    2 -> {
      hero = humanHero {
        name = "Han"
      }
   }
}
```

**Comment by martinbonnin** [comment]
Fixed in 3.1.0
